modules/phase_field/src/ics/BimodalSuperellipsoidsIC.C

```diff
@@ -0,0 +1,243 @@
+  params.addRequiredParam<Real>("small_n", "Exponent n for the randomly placed (small) superellipsoids");
+  params.addParam<Real>("size_variation", 0.0, "Plus or minus fraction of random variation in the semiaxes for uniform, standard deviation for normal");
+  MooseEnum rand_options("uniform normal none","none");
+  params.addParam<MooseEnum>("size_variation_type", rand_options, "Type of distribution that random semiaxes will follow");
+  params.addParam<unsigned int>("numtries", 1000, "The number of tries to place the random particles");
+  return params;
+}
+
+BimodalSuperellipsoidsIC::BimodalSuperellipsoidsIC(const InputParameters & parameters) :
+    SmoothSuperellipsoidBaseIC(parameters),
+    _x_positions(getParam<std::vector<Real> >("x_positions")),
+    _y_positions(getParam<std::vector<Real> >("y_positions")),
+    _z_positions(getParam<std::vector<Real> >("z_positions")),
+    _input_as(getParam<std::vector<Real> >("as")),
+    _input_bs(getParam<std::vector<Real> >("bs")),
+    _input_cs(getParam<std::vector<Real> >("cs")),
+    _input_ns(getParam<std::vector<Real> >("ns")),
+    _npart(getParam<unsigned int>("npart")),
+    _small_spac(getParam<Real>("small_spac")),
+    _large_spac(getParam<Real>("large_spac")),
+    _small_a(getParam<Real>("small_a")),
+    _small_b(getParam<Real>("small_b")),
+    _small_c(getParam<Real>("small_c")),
+    _small_n(getParam<Real>("small_n")),
+    _size_variation(getParam<Real>("size_variation")),
+    _size_variation_type(getParam<MooseEnum>("size_variation_type")),
+    _numtries(getParam<unsigned int>("numtries"))
+{
+}
+
+void
+BimodalSuperellipsoidsIC::initialSetup()
+{
+
+  //Set up domain bounds with mesh tools
+  for (unsigned int i = 0; i < LIBMESH_DIM; i++)
+  {
+    _bottom_left(i) = _mesh.getMinInDimension(i);
+    _top_right(i) = _mesh.getMaxInDimension(i);
+  }
+  _range = _top_right - _bottom_left;
+
+  switch (_size_variation_type)
+  {
+  case 2: //No variation
+    if (_size_variation > 0.0)
+      mooseError("If size_variation > 0.0, you must pass in a size_variation_type in BimodalSuperellipsoidsIC");
+    break;
+  }
+
+  SmoothSuperellipsoidBaseIC::initialSetup();
+}
+
+void
+BimodalSuperellipsoidsIC::computeSuperellipsoidSemiaxes()
+{
+  _as.resize(_input_as.size() + _npart);
+  _bs.resize(_input_bs.size() + _npart);
+  _cs.resize(_input_cs.size() + _npart);
+
+  //First fill in the specified (large) superellipsoids from the input file
+  for (unsigned int i = 0; i < _input_as.size(); ++i)
+  {
+    _as[i] = _input_as[i];
+    _bs[i] = _input_bs[i];
+    _cs[i] = _input_cs[i];
+  }
+
+  //Then fill in the randomly positioned (small) superellipsoids
+  for (unsigned int i = _input_as.size(); i < _input_as.size() + _npart; i++)
+  {
+    //Vary semiaxes
+    switch (_size_variation_type)
+    {
+    case 0: //Random distrubtion, maintaining constant shape
+    {
+      Real rand_num = _random.rand(_tid);
```

Comment:
That may be fine for a random initial condition. If it doesn't produce consistent thread results though just make sure you restrict it to single thread in the test specification so that when we do run with threads in the module we won't have to figure out why this one doesn't work.